File: baekjoon/basic2/9012/9012.cpp
// 백준 9012 : 괄호 (https://www.acmicpc.net/problem/9012)
#include <iostream>
#include <stack>
using namespace std;
int tc;
string s;

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    
    cin >> tc;
    while(tc--){
        stack<char> stk;
        cin >> s;
        stk.push(s[0]);
        for(int i=1; i<s.size(); i++){
            if(stk.size() && stk.top() == '(' && s[i] == ')') stk.pop();
            else stk.push(s[i]);
        }
        if(stk.size()) cout << "NO\n";
        else cout << "YES\n";
    }
    
    return 0;
}

File: baekjoon/basic2/9012/9012.py
# 백준 9012 : 괄호 (https://www.acmicpc.net/problem/9012)
import sys
    
tc = int(sys.stdin.readline())
for _ in range(tc):
    stk = []
    s = sys.stdin.readline().rstrip()
    stk.append(s[0])
    for i in range(1, len(s)):
        if len(stk) and stk[-1] == '(' and s[i] == ')':
            stk.pop()
        else:
            stk.append(s[i])
    if len(stk):
        sys.stdout.write("NO\n")
    else:
        sys.stdout.write("YES\n")